3-explicit-polymorphism/3-explicit-polymorphism.md

```diff
@@ -0,0 +1,156 @@
+# Introduction
+
+Currently in Hazel every function has to be assigned a monomorphic type.
+This makes it impossible to write code that operates on more than one type
+of data. We currently have an example that implements `map : (Num -> Num)
+-> [Num] -> [Num]`. By adopting this proposal, Hazel's map function can
+have the type `forall a. forall b. (a -> b) -> [a] -> [b]`, and thereby be useful for
+a larger set of values.
+
+Specifically, we propose extending Hazel with two type constructs
+and two expression constructs, bringing Hazel's type system up to
+a System F with holes than a Simply Typed Lambda Calculus with holes.
+
+# Type Language
+
+We'd have to extend UHTyp.t with `TyVar` and `Forall` cases:
+
+  ```reason
+  /* UHTyp.re */
+
+  type operand =
+  ...
+  | Forall(TPat.t, t);
+  | TyVar(VarErrStatus.t, Var.t)
+  ```
+
+  A type pattern is the part of the language between the `forall` and the `.`
+  in `forall a. <type>`. Type patterns will be very simple to begin with:
+
+  ```reason
+  /* TPat.t */
+  type t =
+  | Hole(MetaVar.t)
+  | Var(Var.t)
+  ```
+
+  <!-- TODO: Talk about how TPat.t's might expand in the future -->
+
+# Expression Language
+
+  ```reason
+  /* UHExp.t */
+  type operand =
+  ...
+  | TyLam(TPat.t, block)
+  | TyArg(UHTyp.t)
+  ```
+
+# Construct actions
+
+  We'll be adding the following to `Construct(shape)`:
+
+  ```reason
+  type shape =
+  ...
+  | SForall
+  | STyArg
+  ```
+
+  `Construct(SForall)` will be triggered when the user types "forall" followed by
+  a space in a type hole. This will add a Forall type with the cursor at the type
+  pattern hole and a hole as the body.
+
+  `Construct(STyArg)` will be triggered when the user types "type" followed by a
```

Comment:
need a little more detail here -- what happens when there is an expression next to it? are things parenthesized?